src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                if (_notifyFilters.HasFlag(NotifyFilters.Attributes | NotifyFilters.CreationTime | NotifyFilters.LastAccess | NotifyFilters.LastWrite | NotifyFilters.Size))
+                {
+                    _filterFlags = Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemInodeMetaMod | Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemFinderInfoMod;
+                }
+                if (_notifyFilters.HasFlag(NotifyFilters.Security))
+                {
+                    _filterFlags |= Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemChangeOwner | Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemXattrMod;
+                }
+                if (_notifyFilters.HasFlag(NotifyFilters.DirectoryName))
+                {
+                    _filterFlags |= Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsDir | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+                }
+                if (_notifyFilters.HasFlag(NotifyFilters.FileName))
+                {
+                    _filterFlags |= Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+                }
+
+                // The directory has not changed then just restart receiving events. If the directory has changed 
+                // then we need to cleanup the old stream and kick off a new one
+                if (_lastWatchedDirectory.Equals(_directory, StringComparison.CurrentCultureIgnoreCase))
+                {
+                    RestartStream();
+                }
+                else
+                {
+                    CleanupStreamAndWatcher();
+                    CreateStreamAndStartWatcher();
+                    _lastWatchedDirectory = _directory;
```

Comment:
We could save some complexity but it seemed like it was worth a little more complexity to prevent creating a new thread every time a caller stops and starts the watcher. If we think that this isn't worth it, I can definitely remove the reuse code without a problem.

Thoughts?